librustzcash/src/rustzcash.rs

```diff
@@ -1173,3 +1175,196 @@ pub extern "C" fn librustzcash_zip32_xfvk_address(
+}
+
+#[no_mangle]
+pub extern "system" fn librustzcash_mmr_append(
+    // Consensus branch id
+    cbranch: u32,
+    // Length of tree in array representation
+    t_len: u32,
+    // Indices of provided tree nodes, length of p_len
+    ni_ptr: *const u32,
+    // Provided tree nodes data, length of p_len
+    n_ptr: *const [c_uchar; zcash_mmr::MAX_ENTRY_SIZE],
+    // Peaks count
+    p_len: size_t,
+    // New node pointer
+    nn_ptr: *const [u8; zcash_mmr::MAX_NODE_DATA_SIZE],
+    // Return of root commitment (32 byte hash)
+    rt_ret: *mut u8,
+    // Return buffer for appended leaves, should be pre-allocated of log2(t_len)+1 length
+    buf_ret: *mut [c_uchar; zcash_mmr::MAX_NODE_DATA_SIZE],
+) -> u32 {
+    let new_node_bytes: &[u8; zcash_mmr::MAX_NODE_DATA_SIZE] = unsafe {
+        match nn_ptr.as_ref() {
+            Some(r) => r,
+            None => {
+                return 0;
+            } // Null pointer passed, error
+        }
+    };
+
+    let mut tree = match construct_mmr_tree(cbranch, t_len, ni_ptr, n_ptr, p_len, 0) {
+        Ok(t) => t,
+        _ => {
+            return 0;
+        } // error
+    };
+
+    let node = match MMRNodeData::from_bytes(cbranch, &new_node_bytes[..]) {
+        Ok(node) => node,
+        _ => {
+            return 0;
+        } // error
+    };
+
+    let appended = match tree.append_leaf(node) {
+        Ok(appended) => appended,
+        _ => {
+            return 0;
+        }
+    };
+
+    let return_count = appended.len();
+
+    let root_node = tree
+        .root_node()
+        .expect("Just added, should resolve always; qed");
+    unsafe {
+        slice::from_raw_parts_mut(rt_ret, 32).copy_from_slice(&root_node.data().subtree_commitment);
+
+        for (idx, next_buf) in slice::from_raw_parts_mut(buf_ret, return_count as usize)
+            .iter_mut()
+            .enumerate()
+        {
+            tree.resolve_link(appended[idx])
+                .expect("This was generated by the tree and thus resolvable; qed")
+                .data()
+                .write(&mut &mut next_buf[..])
+                .expect("Write using cursor with enough buffer size cannot fail; qed");
+        }
+    }
+
+    return_count as u32
+}
+
+#[no_mangle]
+pub extern "system" fn librustzcash_mmr_delete(
+    // Consensus branch id
+    cbranch: u32,
+    // Length of tree in array representation
+    t_len: u32,
+    // Indices of provided tree nodes, length of p_len+e_len
+    ni_ptr: *const u32,
+    // Provided tree nodes data, length of p_len+e_len
+    n_ptr: *const [c_uchar; zcash_mmr::MAX_ENTRY_SIZE],
+    // Peaks count
+    p_len: size_t,
+    // Extra nodes loaded (for deletion) count
+    e_len: size_t,
+    // Return of root commitment (32 byte hash)
+    rt_ret: *mut u8,
+) -> u32 {
+    let mut tree = match construct_mmr_tree(cbranch, t_len, ni_ptr, n_ptr, p_len, e_len) {
+        Ok(t) => t,
+        _ => {
+            return 0;
+        } // error
+    };
```

Comment:
This can be written more concisely using `.unwrap_or(0)` (and similarly below). https://doc.rust-lang.org/std/result/enum.Result.html#method.unwrap_or

Reply:
Not really, here we're trying to return out of the function. `unwrap_or` is for substituting with a value (e.g. a tree instance) when one does not exist.

Reply:
Oh sorry yes.